File: backend/api/models/payslip.py
"""
Request/response schemas for payslip persistence. `ciphertext_b64`/`iv_b64`
are the AES-256-GCM blob + nonce produced client-side (§4), base64-encoded
for JSON transport — this file, like db/models.py, never carries a
plaintext salary figure.
"""

from pydantic import BaseModel


class PayslipSaveRequest(BaseModel):
    month: str  # e.g. "2026-07"
    ciphertext_b64: str
    iv_b64: str


class PayslipSnapshotOut(BaseModel):
    id: str
    month: str
    created_at: str


class PayslipSnapshotFull(BaseModel):
    """GET /payslip/snapshots — unlike PayslipSnapshotOut (returned from
    POST /save, just a receipt), this carries the ciphertext back so the
    client can decrypt every saved month and compute real trends from them
    (see backend/payslip_trends.py)."""

    id: str
    month: str
    ciphertext_b64: str
    iv_b64: str
    created_at: str


class SessionSummaryOut(BaseModel):
    id: str
    ciphertext_b64: str
    iv_b64: str
    created_at: str


class SessionSummarySaveRequest(BaseModel):
    """Body for POST /payslip/session-summary — the client encrypts the
    plaintext dict POST /chat/summarize returned before sending it here;
    same ciphertext-only contract as PayslipSaveRequest."""

    ciphertext_b64: str
    iv_b64: str


class PayslipParseRequest(BaseModel):
    """Body for POST /payslip/parse — text already extracted client-side
    from an uploaded PDF (frontend/.../PDFParser.tsx). Plaintext; the PDF
    itself never reaches the server."""

    text: str


class PayslipParseResponse(BaseModel):
    fields: dict

File: frontend/src/api/payslip.ts
import { apiClient } from "./client";
import type { EncryptedBlob } from "../crypto/clientEncryption";

export interface PayslipSaveResult {
  id: string;
  month: string;
  created_at: string;
}

/** Throws with `.response.status === 409` (see isDuplicatePayslipError below)
 * if a payslip for this month is already saved — the server checks by
 * month, the one payslip field it ever sees in plaintext (§4). */
export async function savePayslip(month: string, blob: EncryptedBlob): Promise<PayslipSaveResult> {
  const { data } = await apiClient.post<PayslipSaveResult>("/payslip/save", {
    month,
    ciphertext_b64: blob.ciphertextB64,
    iv_b64: blob.ivB64,
  });
  return data;
}

/** True for the "a payslip for this month is already saved" conflict —
 * callers use this to show an informative "already saved" state instead of
 * a red error. */
export function isDuplicatePayslipError(err: unknown): boolean {
  return (err as { response?: { status?: number } })?.response?.status === 409;
}

/** Deletes one saved snapshot — used by the Payslip history panel's
 * per-row delete button and its "Remove duplicates" cleanup action.
 * Deliberately not callable from chat/agents (§ security note in
 * backend/api/routes/payslip.py) — only this explicit, user-clicked path. */
export async function deleteSnapshot(id: string): Promise<void> {
  await apiClient.delete(`/payslip/snapshots/${id}`);
}

export interface SessionSummaryRow {
  id: string;
  ciphertext_b64: string;
  iv_b64: string;
  created_at: string;
}

/** Still ciphertext at this point — decrypt with crypto/clientEncryption.ts before use. */
export async function fetchHistory(): Promise<SessionSummaryRow[]> {
  const { data } = await apiClient.get<SessionSummaryRow[]>("/payslip/history");
  return data;
}

/** Persists an already-encrypted Level 2 summary (see api/chat.ts's summarizeSession). */
export async function saveSessionSummary(blob: EncryptedBlob): Promise<void> {
  await apiClient.post("/payslip/session-summary", {
    ciphertext_b64: blob.ciphertextB64,
    iv_b64: blob.ivB64,
  });
}

/**
 * Turns PDF-extracted text (see components/PayslipUploader/PDFParser.tsx)
 * into structured fields matching ManualEntryForm's shape. The PDF itself
 * never reaches the server — only the text already extracted client-side.
 */
export async function parsePayslipText(text: string): Promise<Record<string, unknown>> {
  const { data } = await apiClient.post<{ fields: Record<string, unknown> }>("/payslip/parse", {
    text,
  });
  return data.fields;
}

export interface PayslipSnapshotRow {
  id: string;
  month: string;
  ciphertext_b64: string;
  iv_b64: string;
  created_at: string;
}

/**
 * Every saved payslip snapshot, ciphertext included — NOT the same as
 * fetchHistory() above, which (despite the similar-sounding name) returns
 * compressed session summaries, not payslip snapshots. Decrypt each row
 * with crypto/clientEncryption.ts before use.
 */
export async function fetchSnapshots(): Promise<PayslipSnapshotRow[]> {
  const { data } = await apiClient.get<PayslipSnapshotRow[]>("/payslip/snapshots");
  return data;
}
